File: java/src/main/java/com/sweetmanor/algorithm/graph/WeightedEdge.java
package com.sweetmanor.algorithm.graph;

import java.util.Objects;

public class WeightedEdge extends Edge implements Comparable<WeightedEdge> {

    public final double weight;

    public WeightedEdge(int u, int v, double weight) {
        super(u, v);
        this.weight = weight;
    }

    @Override
    public WeightedEdge reversed() {
        return new WeightedEdge(v, u, weight);
    }

    @Override
    public int compareTo(WeightedEdge other) {
        Double mine = weight;
        Double theirs = other.weight;
        return mine.compareTo(theirs);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null || getClass() != o.getClass())
            return false;
        WeightedEdge edge = (WeightedEdge) o;
        return Double.compare(weight, edge.weight) == 0;
    }

    @Override
    public int hashCode() {
        return Objects.hash(weight);
    }

    @Override
    public String toString() {
        return u + " " + weight + " -> " + v;
    }

}

File: python/src/com/sweetmanor/algorithm/graph/weighted_edge.py
# 《算法精粹》

from dataclasses import dataclass

from edge import Edge


@dataclass
class WeightedEdge(Edge):
    weight: float

    def reversed(self) -> 'WeightedEdge':
        return WeightedEdge(self.v, self.u, self.weight)

    def __lt__(self, other: 'WeightedEdge') -> bool:
        return self.weight < other.weight

    def __str__(self):
        return f"{self.u} {self.weight} -> {self.v}"
